### PR TITLE
feat(trainer): log epoch and final metrics

### PR DESCRIPTION
**Pull request by borisdayma**
This PR adds new logging capabilities to `Trainer`:

* logs epoch -> this let us compare runs with different batch sizes, using "epoch" as the x-axis instead of "global_step"
* logs final metrics -> metrics are also logged by `Trainer.evaluate` at the end of the script (or at the start if we just use `--do_eval` and not train).

In below example, we see `epoch` has been used for x-axis and `eval_loss` has one extra point corresponding to the end of the script.
image

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -166,6 +166,8 @@ def __init__(
             # Set an xla_device flag on the model's config.
             # We'll find a more elegant and not need to do this in the future.
             self.model.config.xla_device = True
+        self.global_step = 0
+        self.epoch = 0
```
Comment (borisdayma) [inline src/transformers/trainer.py:170]: We keep track of `global_step` and `epoch` so they can be accessible by `evaluate` method which is called at the end of the scripts.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -398,53 +399,52 @@ def train(self, model_path: Optional[str] = None):
 
                     scheduler.step()
                     model.zero_grad()
-                    global_step += 1
+                    self.global_step += 1
+                    self.epoch = epoch + (step + 1) / len(epoch_iterator)
```
Comment (borisdayma) [inline src/transformers/trainer.py:403]: We keep track of `epoch` so we can use it as an x-axis on comparative charts. Also it's calculated at each step so we have the right value in case we run `Trainer.evaluate` method outside of `Trainer.train` method.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -398,53 +399,52 @@ def train(self, model_path: Optional[str] = None):
...
                             logs["learning_rate"] = learning_rate_scalar
                             logs["loss"] = loss_scalar
                             logging_loss = tr_loss
 
-                            if self.tb_writer:
-                                for k, v in logs.items():
-                                    self.tb_writer.add_scalar(k, v, global_step)
-                            if is_wandb_available():
-                                wandb.log(logs, step=global_step)
+                            self._log(logs)
```
Comment (borisdayma) [inline src/transformers/trainer.py:416]: This is where we log data to tensorboard and wandb.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -398,53 +399,52 @@ def train(self, model_path: Optional[str] = None):
...
+
+                            if self.args.evaluate_during_training:
+                                results = self.evaluate()
+                                for key, value in results.items():
+                                    eval_key = "eval_{}".format(key)
+                                    logs[eval_key] = value
```
Comment (borisdayma) [inline src/transformers/trainer.py:422]: We moved this section here to avoid logging twice the same data to tensorboard and wandb (which would not be an issue but is useless) as the data is logged directly within `Trainer.evaluate`.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -455,7 +455,15 @@ def train(self, model_path: Optional[str] = None):
             self.tb_writer.close()
 
         logger.info("\n\nTraining completed. Do not forget to share your model on huggingface.co/models =)\n\n")
-        return TrainOutput(global_step, tr_loss / global_step)
+        return TrainOutput(self.global_step, tr_loss / self.global_step)
+
+    def _log(self, logs: Dict[str, torch.Tensor]) -> None:
```
Comment (borisdayma) [inline src/transformers/trainer.py:460]: We created a separate function as we now log in 2 places:
* training data in `train`
* evaluation data in `evaluate`

This is also where we add `epoch`.

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -572,6 +580,12 @@ def evaluate(
 
         output = self._prediction_loop(eval_dataloader, description="Evaluation")
 
+        logs = {}
+        for key, value in output.metrics.items():
+            eval_key = "eval_{}".format(key)
+            logs[eval_key] = value
+        self._log(logs)
```
Comment (borisdayma) [inline src/transformers/trainer.py:587]: Evaluation metrics are now directly logged in `evaluate` so that we have access to final model metrics at the end of the run.

**Review by borisdayma — COMMENTED** [review]
I added a few comments on the code to explain the changes.

**Comment by borisdayma** [comment]
Already merged through PR #4324
